pkg/scheduler/queue/tree_queueing_algorithms.go

```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package queue
+
+// QueuingAlgorithm represents the set of operations specific to different approaches to queuing/dequeuing. It is
+// applied at the layer-level -- every Node at the same depth in a MultiQueuingAlgorithmTreeQueue shares the same QueuingAlgorithm,
+// including any state in structs that implement QueuingAlgorithm.
+type QueuingAlgorithm interface {
+	// addChildNode creates a child Node of parent. QueuePaths passed to enqueue functions are allowed to contain
+	// node names for nodes which do not yet exist; in those cases, we create the missing nodes. Implementers are
+	// responsible for adding the newly-created node to whatever state(s) keep track of nodes in the subtree, as well
+	// as updating any positional pointers that have been shifted by the addition of a new node.
+	addChildNode(parent, child *Node)
+
+	// dequeueSelectNode, given a node, returns the node which an element will be dequeued from next, as well
+	// as a bool representing whether the passed node and all its children have been checked for dequeue-ability.
+	// The returned node may be the same as the node passed in; this triggers the base case for the recursive
+	// Node.dequeue, and results in dequeuing from the node's local queue. dequeueSelectNode does *not*
+	// update any Node fields.
```

Comment:
I'm looking to express somehow that the algo needs to maintain internal state about whether all nodes have been checked, and why, and a hint that it will need to be reset eventually.

Take or leave what you want from here:
```
// dequeueSelectNode selects either itself or one of its child Nodes to be dequeued from next.
// Implementations are expected to select the next node based on data directly available
// in the Node or in the QueuingAlgorithm's internal state, not by traversing any child Nodes.
// If a child selected by dequeueSelectNode cannot be dequeued from, the Tree will return
// back to this Node and call dequeueSelectNode node again to check the next option.
// For this reason, dequeueSelectNode must also return a boolean to indicate whether
// all possible self and child Node options have been checked within this dequeue cycle.
// The QueuingAlgorithm's internal tracking for whether all options have been checked should be reset upon dequeue
// ...etc about the base case
```